dataframe.py

```diff
@@ -0,0 +1,270 @@
+    @abstractmethod
+    def type(self) -> DataType:
+        """
+        Return the logical type of each column cell value
+        """
+        pass
+
+    def to_numpy(self):
+        """
+        Access column's data as a NumPy array. Recommended to return a view if
+        able but not required
+        """
+        raise NotImplementedError("Conversion to NumPy not available")
+
+    def to_arrow(self, **kwargs):
+        """
+        Access column's data in the Apache Arrow format as pyarrow.Array or
+        ChunkedArray. Recommended to return a view if able but not required
+        """
+        raise NotImplementedError("Conversion to Arrow not available")
+
+
+# ----------------------------------------------------------------------
+# DataFrame: the main public API
+
+
+class DataFrame(ABC, Mapping):
+    """
+    An abstract data frame base class.
+
+    A "data frame" represents an ordered collection of named columns. A
+    column's "name" is permitted to be any hashable Python value, but strings
+    are common. Names are not required to be unique. Columns may be accessed by
+    name (when the name is unique) or by position.
+    """
+
+    def __dataframe__(self):
+        """
+        Idempotence of data frame protocol
+        """
+        return self
+
+    def __iter__(self):
+        # TBD: Decide what iterating should return
+        return iter(self.column_names)
+
+    def __len__(self):
+        return self.num_rows
+
+    @property
+    @abstractmethod
+    def num_columns(self):
+        """
+        Return the number of columns in the DataFrame
+        """
+        pass
+
+    @property
+    @abstractmethod
+    def num_rows(self):
+        """
+        Return the number of rows in the DataFrame (if known)
+        """
+        pass
+
+    @abstractmethod
+    def iter_column_names(self) -> Iterable[Any]:
+        """
+        Return the column names as an iterable
+        """
+        pass
+
+    # TODO: Should this be a method or property?
+    @property
+    @abstractmethod
+    def column_names(self) -> Sequence[Any]:
+        """
+        Return the column names as a materialized sequence
+        """
+        pass
+
+    # TODO: Should this be a method or property?
+    @property
+    def row_names(self) -> Sequence[Any]:
+        """
+        Return the row names (if any) as a materialized sequence. It is not
+        necessary to implement this method
+        """
+        raise NotImplementedError("row_names")
+
+    def __getitem__(self, key: Hashable) -> Column:
+        return self.column_by_name(key)
+
+    @abstractmethod
+    def column_by_name(self, key: Hashable) -> Column:
+        """
+        Return the column whose name is the indicated key
+        """
+        pass
+
+    @abstractmethod
+    def column_by_index(self, i: int) -> Column:
+        """
+        Return the column at the indicated position
+        """
+        pass
+
+
+class MutableDataFrame(DataFrame, MutableMapping):
```

Comment:
I also don't think that mutability makes sense in an interchange API like this. I put this here to "stir the pot" and elicit reactions from others.

Reply:
I think when it comes to performance/memory usage, it's important for the users/libs to be able to modify values inplace. But I guess you could argue we could convert everything first to numpy arrays and then deal with that.

Reply:
The main problem I see with this is that an implementer of the protocol/API may have to deal with converting from foreign input. For example:

```
df[col] = data
```

so if `data` is a NumPy array and `df` doesn't use NumPy arrays internally, then it's a can of worms to account for all of the different potential kinds of `data` that the user may pass in

Reply:
I think another downfall of inplace is that it is often deceptive and leaky. In pandas, the overwhelming majority of the time it is not really inplace, it is copying at some point, so users who think they are being more efficient or using best practices are often not getting the desired effect under the hood. Being functional (`df.<op>` returns another dataframe) is better because external libraries should ideally not have to think about optimizing calls for performance or memory, but rather trust that those who implemented the library did a good job.

Reply:
That may be true on pandas, but some simple arithmetic operations on numpy show a different story, for isntance, if `a`, and `b` are random `float64` arrays is `10000x10000`, we have:

```
In [12]: %timeit np.add(a, b, out=a)                                                                                                                                                                                                   
116 ms ± 2.8 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [13]: %timeit np.add(a, b)                                                                                                                                                                                                          
483 ms ± 200 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

This may also just be indicative of us preferably having to be operating on a more low level layer (like numpy) rather than the data frame implementations, if we want to do arithmetics.

Reply:
There's one more thing to consider I did not mention before: From a library perspective, editing input data inplace is very destructive to the user's environment. 

Often there is a disconnect between users and those who maintain or create libraries. You will hear and see a lot of arguments about performance, but in reality users just do not care as much about performance as they do other things. It is easy for a library maintainer to focus on performance, it is a raw number we can try to optimize. However the user cares more about their own productivity and about prediction accuracy. A real world difference of a few seconds or even minutes is noise to a user who is interpreting results. 

Also minor note: Your `timeit` example does not make sense in practice because each time the `In[12]` trial updates `a`. The original `a` values are gone after the first iteration. It's a minor point because it doesn't actually affect performance. The outputs of `In[12]` and `In[13]` are not the same.

Ignoring this correctness issue for a moment, you can improve the performance of the `In[13]` line by almost a factor of 2 if you do the following (provided you would want to store the output in some variable `c`):

```
%%timeit c = np.copy(a); np.add(a, b, out=c)
```

This gets to my point about performance. Optimizing for the behaviors of some other library that is out of your control is a dangerous business to be in, especially when you destroy a user's environment to do so. If you are the user destroying your own environment, that's perfectly fine, but the purpose of this interface is for libraries to avoid depending directly on pandas (forcing conversion into pandas). From a uniform interface perspective, we want to discourage in place updates because it is a leaky abstraction (should dataframe consumers really know about or rely on internal state?) and destructive to the user's environment. 

Users may think they want this, but the same thing can be functionally achieved by `df = df.update_inplace()`. The only argument to keep it is a performance argument, and I feel this is a very weak argument. If a dataframe library is not effectively managing memory or has performance issues that is an issue the dataframe library should fix, not something to be attempted to be worked around by manually changing internal state in place.

Reply:
The performance is something we really care about in sklearn. We may not always be optimal, but we try, and when we do change things in place, it's not the user's data we're changing. It's an intermediate array we have in our methods. The optimizations we do even touch the level of an array being f-continuous or c-continuous, depending on the arithmatics we need to apply to the data. So at least for our usecase, these details really do matter.